Refactor: convert_hf_to_gguf.py

pwilkin opened the pull request:
From the creators of the horror show "llama-model.cpp - refactoring" comes the sequel: "convert_hf_to_gguf.py - refactoring".

Still in draft because it's not all finished, but I thought that if someone wanted to chime in I'll put it up already.

CISC commented:
:exploding_head: Looks like a decent enough plan though. Only downside is that there will no longer just be a single script to grab and run, but that has been a source of error and confusion anyway.

You should derive `_MODEL_MODULES` from `models_map`.

compilade commented on conversion/__init__.py:
```diff
@@ -0,0 +1,191 @@
+    }
+    if name not in model_map:
+        raise ValueError(f"Unknown model class: {name}")
+    module_name, class_name = model_map[name]
```
Since the `class_name` seems to always be equal to `name`, would it make sense to lighten the `model_map` to avoid having to write the class name twice every time?

pwilkin replied:
Yup.

compilade commented:
(note for later) This will (trivially) conflict with at least 

- #17069
- #15667
- #15727
- (non-existent yet, but wip) convert : generalized repacking for pre-quantized models

CISC commented:
> - #17069
> - #15667

Are you expecting more work on these? Can be merged now.

pwilkin commented:
Okay, so I've actually gone over and verified the conversion results - refactored, applied some tips from here while doing that and most of all - tested that it actually works.

pwilkin commented:
Conversion verification script: https://gist.github.com/pwilkin/1e488423e9f2549c0518179bb9f752d5

pwilkin committed:
I'm baaaaack! Did you miss me?

pwilkin committed:
Fix linter errors

pwilkin committed:
And these 2

pwilkin commented:
@ngxson shamed me by finishing the model code refactor, so I decided to come back to this oldie-but-goodie.

@CISC something for ya :)

pwilkin commented:
P.S. The `analyze_conversion_faithfulness.py` script is of course temporary.

CISC commented on conversion/__init__.py:
```diff
@@ -0,0 +1,322 @@
+
+
+TEXT_MODEL_MAP: dict[str, tuple[str, str]] = {
+    "AfmoeForCausalLM": ("afmoe", "AfmoeModel"),
```
What is the purpose of this tuple again? The second item seems to be unused.

CISC commented on convert_hf_to_gguf.py:
```diff
@@ -13474,4 +258,6 @@ def main() -> None:
 
 
 if __name__ == '__main__':
+    # Ensure all model modules are imported so their @ModelBase.register decorators run.
+    load_all_models()
```
```suggestion
```
Why are we doing that? The correct one is dynamically loaded.

CISC commented on convert_lora_to_gguf.py:
```diff
@@ -22,11 +22,13 @@
+from conversion import LazyTorchTensor, ModelBase, load_all_models
 
-from gguf.constants import GGUFValueType
+# Ensure all model classes are registered before we use them.
+load_all_models()
```
```suggestion
```
Again, should not be necessary, use `get_model_class` instead of `ModelBase.from_model_architecture` below.

CISC commented:
@pwilkin Mind focusing a little bit on this one this week so we can bring it to completion?

pwilkin commented:
@CISC yup, will take a look tomorrow or Wed probably (unless I'm still alive in the evening, then maybe today).

pwilkin committed:
integrate code review comments

pwilkin commented:
@CISC alrighy, should be good to go.

pwilkin committed:
Suppress ty mistral errors

CISC commented:
I'd like to merge #22597 first

CISC commented:
@pwilkin Sigh, sorry for the delay yet again, it has been hectic lately, but given it will be a long weekend starting tomorrow I'm hoping we can wrap this up this week if you have the time.

pwilkin commented:
Aight, let me rebase again.

pwilkin committed:
Integrate upstream convert_hf_to_gguf.py changes into split modules

Ports 15 upstream commits (05e141a6b..5d44db600) that touched the
monolithic convert_hf_to_gguf.py into the new conversion/*.py layout
introduced by the refactor split.

New text/mmproj architectures registered:
  GraniteSpeechForConditionalGeneration, MiMoV2ForCausalLM,
  MiniCPMV4_6ForConditionalGeneration, Sarashina2VisionForCausalLM,
  SarvamMoEForCausalLM (+ modeling_sarvam_moe.SarvamMoEForCausalLM).

Notable changes:
- filter_tensors classmethod added to ModelBase/TextModel/MmprojModel
  and wired into index_tensors; many model classes refactored to move
  tensor-name skip/rename logic out of modify_tensors and into
  filter_tensors (upstream #22597).
- LlamaModel._repack_nvfp4 override (Q/K RoPE permutation, #22611).
- MistralModel yarn apply_scale support (#22612).
- Gemma4Model._generate_nvfp4_tensors override for 26B NVFP4 (#22804).
- LlavaVisionModel image-break token fallback for Mistral params.json
  -1 placeholders (#22914).
- Pixtral 12B --mistral-format conversion fixes (#22981).
- FP8 KV-cache scales fix (#22818) and uint dtype byteswap disable
  (#18908).

New files:
  conversion/sarashina2.py (Sarashina2VL text + vision)

CISC commented:
Ah, quick note, can you rename the files similarly to `src/models/`, ie. `bailingmoe.py`, `kimi-linear.py`, etc?

CISC commented on conversion/granite.py:
```diff
@@ -0,0 +1,330 @@
+
+    def modify_tensors(
+        self, data_torch: Tensor, name: str, bid: int | None
+    ) -> Iterable[tuple[str, Tensor]]:
```
```suggestion
    def modify_tensors(self, data_torch: Tensor, name: str, bid: int | None) -> Iterable[tuple[str, Tensor]]:
```

CISC replied:
I kept the analyze script BTW, it was the one pointing this out. :)

CISC reviewed (approved):
Do the renames and we should be ready to merge.

pwilkin committed:
Update conversion/granite.py

Co-authored-by: Sigbjørn Skjæret <sigbjorn.skjaeret@scala.com>

pwilkin committed:
Renames

pwilkin commented:
@CISC aight, renames done.

CISC commented on conversion/grovemoe.py:
`grovemoe`

CISC commented:
@ggml-org/maintainers Need a second approval

ServeurpersoCom commented:
Wow, that's amazing!!! I didn't expect to see that so soon, 14K lines of Python refactored!

CISC commented:
> Wow, that's amazing!!! I didn't expect to see that so soon, 14K lines of Python refactored!

..only been half a year... :D

ggerganov commented:
> > Wow, that's amazing!!! I didn't expect to see that so soon, 14K lines of Python refactored!
> 
> ..only been half a year... :D

I was about to ask to wait to merge #22673 first later today, but then I noticed this PR is from Nov last year and felt bad and decided not to ask 😄

CISC commented:
> > > Wow, that's amazing!!! I didn't expect to see that so soon, 14K lines of Python refactored!
> > 
> > 
> > ..only been half a year... :D
> 
> I was about to ask to wait to merge #22673 first later today, but then I noticed this PR is from Nov last year and felt bad and decided not to ask 😄

Whichever is fine, but yeah, this one has been sitting a while, would be nice to finally get it done.

ggerganov reviewed (approved):
The MTP branch should be easy to rebase after this is merged.

CISC commented:
@pwilkin Care to do the final rename, then merge? :)

CISC commented:
Ah, hm, `ty` doesn't like the `.base` etc imports.

pwilkin committed:
grove -> grovemoe

pwilkin commented:
It doesn't because Python module names can't contain dashes, so we can't keep the convention, the filenames have to have underscores instead :)

pwilkin committed:
Rename to underscore format.

CISC commented:
> It doesn't because Python module names can't contain dashes, so we can't keep the convention, the filenames have to have underscores instead :)

Ah, crap. :)

CISC reviewed (approved):
Ok, that fixed it.

fluffle commented:
Slightly suspicious that you need to include the "conversion" dir in all the docker images now? At least, the latest docker images can't run `convert_hf_to_gguf.py` any longer because of `ModuleNotFoundError: No module named 'conversion'`...

CISC commented:
> Slightly suspicious that you need to include the "conversion" dir in all the docker images now? At least, the latest docker images can't run `convert_hf_to_gguf.py` any longer because of `ModuleNotFoundError: No module named 'conversion'`...

Forgot that was a thing, will fix.
